File: app/schemas/vacancy.py
from pydantic import BaseModel, HttpUrl, Field
from typing import Optional, List
from datetime import datetime
from .base import BaseSchema, TimestampMixin
from .company import CompanyInVacancy
from .experience import ExperienceInVacancy
from .work_format import WorkFormatInVacancy
from .work_schedule import WorkScheduleInVacancy
from .skill import SkillInVacancy, VacancySkillCreate

# Базовые схемы
class VacancyBase(BaseSchema):
    title: str
    description: str
    salary_from: Optional[float] = None
    salary_to: Optional[float] = None
    currency: Optional[str] = None
    location: Optional[str] = None
    raw_address: Optional[str] = None
    parsed_address: Optional[str] = None
    source_url: HttpUrl
    published_date: Optional[datetime] = None
    is_active: bool = True

# Создание вакансии (с ID внешних ключей)
class VacancyCreate(VacancyBase):
    company_id: int
    experience_id: int
    work_format_id: int
    work_schedule_id: int
    skills: List[VacancySkillCreate] = Field(default_factory=list)  # Навыки при создании

class VacancyUpdate(BaseModel):
    title: Optional[str] = None
    description: Optional[str] = None
    salary_from: Optional[float] = None
    salary_to: Optional[float] = None
    currency: Optional[str] = None
    location: Optional[str] = None
    is_active: Optional[bool] = None
    experience_id: Optional[int] = None
    work_format_id: Optional[int] = None
    work_schedule_id: Optional[int] = None

# Схемы ответа
class VacancySimple(VacancyBase, TimestampMixin):
    vacancy_id: int
    company_id: int
    experience_id: int
    work_format_id: int
    work_schedule_id: int

# Полная вакансия со связанными объектами
class Vacancy(VacancySimple):
    company: Optional[CompanyInVacancy] = None
    experience: Optional[ExperienceInVacancy] = None
    work_format: Optional[WorkFormatInVacancy] = None
    work_schedule: Optional[WorkScheduleInVacancy] = None
    skills: List[SkillInVacancy] = []

# Вакансия с деталями компании (для списков)
class VacancyWithCompany(VacancySimple):
    company: Optional[CompanyInVacancy] = None
    experience: Optional[ExperienceInVacancy] = None

# Для фильтрации и поиска
class VacancyFilter(BaseModel):
    title: Optional[str] = None
    company_id: Optional[int] = None
    experience_id: Optional[int] = None
    work_format_id: Optional[int] = None
    work_schedule_id: Optional[int] = None
    location: Optional[str] = None
    min_salary: Optional[float] = None
    max_salary: Optional[float] = None
    currency: Optional[str] = None
    skill_ids: Optional[List[int]] = None
    is_active: Optional[bool] = True
    offset: int = 0
    limit: int = Field(100, le=1000)

# Статистика
class VacancyStats(BaseSchema):
    total: int
    active: int
    with_salary: int
    average_salary: Optional[float] = None

File: frontend/src/types/vacancy.ts
export type CompanyInVacancy = {
  company_id: number;
  name: string;
};

export type ExperienceInVacancy = {
  experience_id: number;
  name: string;
};

export type WorkFormatInVacancy = {
  work_format_id: number;
  name: string;
};

export type WorkScheduleInVacancy = {
  work_schedule_id: number;
  name: string;
};

export type SkillInVacancy = {
  skill_id: number;
  name?: string | null;
  category?: string | null;
  is_mandatory: boolean;
};

export type VacancyBase = {
  title: string;
  description: string;
  salary_from?: number | null;
  salary_to?: number | null;
  currency?: string | null;
  location?: string | null;
  raw_address?: string | null;
  parsed_address?: string | null;
  source_url: string;
  published_date?: string | null;
  is_active: boolean;
};

export type VacancySimple = VacancyBase & {
  vacancy_id: number;
  company_id: number;
  experience_id: number;
  work_format_id: number;
  work_schedule_id: number;
};

export type VacancyWithCompany = VacancySimple & {
  company?: CompanyInVacancy | null;
  experience?: ExperienceInVacancy | null;
};

export type Vacancy = VacancySimple & {
  company?: CompanyInVacancy | null;
  experience?: ExperienceInVacancy | null;
  work_format?: WorkFormatInVacancy | null;
  work_schedule?: WorkScheduleInVacancy | null;
  skills: SkillInVacancy[];
};

export type VacancyFilters = {
  title?: string;
  location?: string;
  min_salary?: number;
  max_salary?: number;
  experience_id?: number;
  work_format_id?: number;
  work_schedule_id?: number;
  skill_ids?: number[];
  is_active?: boolean;
  offset?: number;
  limit?: number;
};
